doc/source/reference/thread_safety.rst

```diff
@@ -22,15 +22,30 @@ are not reproducible, let alone correct. It is also possible to crash the Python
 interpreter by, for example, resizing an array while another thread is reading
 from it to compute a ufunc operation.
 
-In the future, we may add locking to ndarray to make writing multithreaded
+In the future, we may add locking to :class:`~numpy.ndarray` to make writing multithreaded
 algorithms using NumPy arrays safer, but for now we suggest focusing on
 read-only access of arrays that are shared between threads, or adding your own
 locking if you need to mutation and multithreading.
 
-Note that operations that *do not* release the GIL will see no performance gains
+Note that operations that *do not* release the :term:`python:GIL` will see no performance gains
 from use of the `threading` module, and instead might be better served with
 `multiprocessing`. In particular, operations on arrays with ``dtype=object`` do
-not release the GIL.
+not release the :term:`python:GIL`.
+
+Thread-local state
```

Comment:
errstate is stored in a contextvar, so it's actually context-local state. It's async-safe and thread-safe.

Maybe you could call it `Context-local state`? I'm not sure if there's a term for it that asyncio or CPython more generally uses.

Also the printoptions are a contextvar as well.

Reply:
The custom allocator object is also context-local: https://numpy.org/devdocs/reference/c-api/data_memory.html#configurable-memory-routines-in-numpy-nep-49